specs/Standards/Wallets/InjectedWallets.md

```diff
@@ -0,0 +1,369 @@
+    [transactions.functionCall(
+      "addMessage",
+      { text: "Hello World!" },
+      utils.format.parseNearAmount("0.00000000003"),
+      utils.format.parseNearAmount("0.01")
+    )],
+    utils.serialize.base_decode(block.header.hash)
+  ),
+});
+
+// Send the transaction to the blockchain.
+await provider.sendTransaction(signedTx);
+```
+
+##### `signTransactions`
+
+Sign a list of transactions. This request should require explicit approval from the user.
+
+```ts
+import { transactions, providers, utils } from "near-api-js";
+import { AccessKeyView } from "near-api-js/lib/providers/provider";
+
+// Retrieve accounts (assuming already connected) and current network.
+const { network, accounts } = window.near.wallet;
+
+// Setup RPC to retrieve transaction-related prerequisites.
+const provider = new providers.JsonRpcProvider({ url: network.nodeUrl });
+
+const [block, accessKey] = await Promise.all([
+  provider.block({ finality: "final" }),
+  provider.query<AccessKeyView>({
+    request_type: "view_access_key",
+    finality: "final",
+    account_id: accounts[0].accountId,
+    public_key: accounts[0].publicKey.toString(),
+  }),
+]);
+
+const signedTxs = await window.near.wallet.signTransactions({
+  transactions: [
+    transactions.createTransaction(
+      accounts[0].accountId,
+      accounts[0].publicKey,
+      "guest-book.testnet",
+      accessKey.nonce + 1,
+      [transactions.functionCall(
+        "addMessage",
+        { text: "Hello World! (1/2)" },
+        utils.format.parseNearAmount("0.00000000003"),
+        utils.format.parseNearAmount("0.01")
+      )],
+      utils.serialize.base_decode(block.header.hash)
+    ),
+    transactions.createTransaction(
+      accounts[0].accountId,
+      accounts[0].publicKey,
+      "guest-book.testnet",
+      accessKey.nonce + 2,
+      [transactions.functionCall(
+        "addMessage",
+        { text: "Hello World! (2/2)" },
+        utils.format.parseNearAmount("0.00000000003"),
+        utils.format.parseNearAmount("0.01")
+      )],
+      utils.serialize.base_decode(block.header.hash)
+    )
+  ]
+});
+
+for (let i = 0; i < signedTxs.length; i += 1) {
+  const signedTx = signedTxs[i];
+  
+  // Send the transaction to the blockchain.
+  await provider.sendTransaction(signedTx);
+}
+```
+
+##### `disconnect`
+
+Remove visibility of all accounts from the wallet.
+
+```ts
+await window.near.wallet.disconnect();
+```
+
+##### `signIn`
+
+Add `FunctionCall` access key(s) for one or more accounts. This request should require explicit approval from the user.
+
+```ts
+import { utils } from "near-api-js";
+
+// Retrieve the list of accounts we have visibility of.
+const { accounts } = window.near.wallet;
+
+// Request FunctionCall access to the 'guest-book.testnet' smart contract for each account.
+await window.near.wallet.signIn({
+  permission: {
+    receiverId: "guest-book.testnet",
+    methodNames: [],
+  },
+  accounts: accounts.map(({ accountId }) => {
```

Comment:
So is it expected that `connect()` and `signIn()` are now separate actions that the user needs to respond to in the wallet ?

It's mentioned that during `connect()` the user must select which accounts they want to "expose" to the dapp. And then the `signIn()` action is separate to allow a `FunctionCall` access on the accounts... This feels like a lot of steps which could be combined into one? (`connect()` could optionally pass `FunctionCall` parameters if the dapp wants method access on the selected accounts)

I also feel it's a little strange connecting with multiple accounts, seems like it might be confusing UX for users and requires more management on the dapp side to make sure account selection is well implemented for user actions. Whenever I've connected to a dapp- its usually with the expectation of a single account.

Reply:
Yes - `connect()` is separate from `signIn()` for a few reasons.

**From the dApp side:**
`connect()` means 'wallet, please tell me what account(s) I can request signing requests for'
`signIn()` means 'wallet, please add this access key to one of the accounts that was exposed when I `connect`'d to the wallet, so that I (the dapp) can submit transactions using that key instead of having to request explicit signing via the wallet even for simple and 'safe' transactions.

**From the wallet side:**
`connect()` means 'dApp, here are the accounts that the user would like to be able to use to interact with you'
`signIn()` means 'dApp, I have added an access key to the requested account so that you can submit limited transactions directly on behalf of the user without needing explicit Wallet interactions for each transaction'

Because `connect()` does not involve a blockchain transaction and is _only_ an interaction between the dApp and the wallet, its failure modes are specific to wallet interactions and will never include failures such as 'transaction failed' etc.

Because `signIn()` involves adding an access key to the requested account, its failure mode is specific to those modes encountered when a blockchain transaction fails for whatever reason.

The separation between `connect()` and `signIn()` becomes more clear when we look forward to off-chain account verification.  

For example, lets say you want to have someone 'log in' to your off-chain forum site by using their NEAR account as their identity.  

Currently, you would need to have the user `signIn()` so that you could verify that they own a particular account by looking on-chain and verifying that they added the access key you requested to the account.  Although this approach is secure (we know that someone owns a particular account if they can add access keys to it), it is sub-optimal in cases where the requesting app never intends to submit **any** transactions using that limited access key; essentially we are having the user mutate their on-chain account to prove they own it, but in a way that costs them on-chain storage and isn't actually useful after they have added the key, thus proving they own the account.

Note that this can be the case even for dApps that do interact with the blockchain! Some dApps require that all transactions they submit must be signed using a Full Access Key, so the 'sign in' process that creates a limited access key is superfluous to them.

So the idea is that while you will always `connect()` a dApp to a wallet instance, depending on the dApp type, it may request a `signIn()` if the dApp needs/wants a limited access key in its local scope, but alternatively it may request something like e.g. `verifyAccountOwner()` instead, if the dApp only needs to be absolutely sure that the requesting user controls a particular on-chain account.  Note that the failure mode of this new `verifyAccountOwner()` method will be different than either `connect()` or `signIn()` because it will not involve submitting an on-chain transaction; by keeping these behaviours separate, we allow for more flexible usage patterns and simplify each of the APIs involved and their possible failure modes.

Reply:
> **From the dApp side:**
> connect() means 'wallet, please tell me what account(s) I can request signing requests for'

> **From the wallet side:**
> connect() means 'dApp, here are the accounts that the user would like to be able to use to interact with you'

I just feel like this is going to be a confusing UX for the user, as well as for wallets- now we have to provide an interface for users to select multiple accounts on a `connect()` request ? I don't think I've ever "logged in" to a website with multiple accounts before and I don't really see the benefit ? Especially since we can easily swap and change between accounts by just making a new connect request to the wallet (potentially there could be a `swapAccount()` method or something of the like, which opens the wallet to select a new account, even if we are currently "logged in" with an account). What I was trying to say is that I feel this is going to cause more complexity for both dapps and wallets (and potentially confuse users). More direct action paths are always better in my view.

Also dapps are now going to have to know for sure which one of the multiple selected wallets the user wanted to complete an action with. Does this mean its expected of all dapps to provide a kind of account dropdown list (or similar UI) for "currently active" account? You wouldn't want to do a forum action with the wrong account by mistake... now that we have "connected" with multiple accounts we require additional layers to prevent such mistakes.

If you are hard set on this interface, then I would request that you allow Dapps to provide an extra parameter to `connect()` - something like `{ occupancy?: "multple" | "single" | number }` - that would allow those Dapps which only expect to work with single accounts the option to build their UI around that.

Lastly, I do hope that the `connect()` method is going to provide even more context to the Wallet. Especially `network`, as I mentioned in another comment. Connecting with a wallet in an incorrect network is only going to get caught later on the Dapp side and could be a surprising error for users. Better to have these things determined from the get go.

Reply:
It does make things more complex to allow users to select more than one account to be shared with the dApp from a wallet -- but this is actually a very often requested feature that we hope to support as a first-class concept in the API.  Note that although `connect()` can return multiple accounts, _only one of those accounts generally would be considered 'active' at once in the dApp context_ -- and generally that account would be set after either a `verifyOwner()` or `signIn()` call has been completed for the account that the user wants to be their _active_ account.

Although each dApp author could write their own code to handle interacting with wallets and prompting the user to select the 'active' account, we don't think it's a good use of their time to do so (this is the equivalent of boiler-plate that distracts the dApp developer from solving the real problems they are trying to solve).  This aligns with the philosophy that integrating with many different wallets and providing the functionality to select which of those wallets to use should also be something that we provide tools for the dApp author to use. We will be adding this functionality to the wallet selector package / modal UIs, so that the dApp developer can use the modal UI it provides to allow the user to see the list of accounts, and to select which account should be the 'active' account for signing in, verifying owner, or signing requests.

If a dApp author wants to write their own account switcher/selector UI, they could do so by piggy-backing on the core package which exposes events, accounts, and a concept of 'active account' for the dApp developer -- or they could just use the UI components that we publish from the wallet-selector and be up-and-running immediately.

Note that it currently stands, we do not dictate that a wallet _must_ return multiple accounts from a `connect()` call, _even if there are multiple loaded into that wallet_.  The only requirement is that it returns an array with _at least_ 1 account in it -- which allows wallet authors to make the UI for the `connect()` method simple (select one) or more complex (select multiple) as they see fit for their wallet design, while ensuring that our API is also compatible with more than one being returned.

@lostpebble Does this address your concerns here, or do you think that adding `occupancy` as an argument to the `connect()` method is still a requirement for this to be reasonable, keeping in mind that if we do add that option, wallet providers implementation and UI will become more complex.

Reply:
Hi @MaximusHaximus - thanks for taking Network into consideration with the `connect()` method.

> @lostpebble Does this address your concerns here, or do you think that adding occupancy as an argument to the connect() method is still a requirement for this to be reasonable, keeping in mind that if we do add that option, wallet providers implementation and UI will become more complex.

I think its okay to leave it out- as long as its always okay for a user to only select a single account to connect with. If you think certain dApps might require the user to connect with multiple (or a certain amount) of accounts, for whatever reason, then having an `occupancy` option might be a good idea so the wallet can show the relevant UI. But I don't think should ever be the case?

This "multi-account" concept does complicate the Wallet UI a bit though. I guess in our case, when a Dapp calls the `connect({ network: "testnet" })` method, we'll show the user an interface that allows them to do one of the following:

* Select a single testnet account
* Select multiple specific testnet accounts
* Connect with all known wallet testnet accounts

We'll have to think of ways of streamlining such an interface.

And after that, I would suppose that future calls to `verifyOwner()` do not require any further prompts from the user (since they've already allowed the connection). But `signIn()` would run the specific sign-in flow for the account the dApp has chosen to use.